### ISSUE TITLE
Heroku Dyno crashes when setting up Payee->Category Database

### ISSUE THREAD

**Issue by CyberiumShadow** [BrodieSutherland/up-to-ynab#2]
```021-10-03T12:34:14.611251+00:00 app[web.1]: Setting up Payee->Category Databases...

2021-10-03T12:34:16.439245+00:00 app[web.1]: Traceback (most recent call last):

2021-10-03T12:34:16.439276+00:00 app[web.1]:   File "/app/main.py", line 33, in <module>

2021-10-03T12:34:16.439367+00:00 app[web.1]:     helper.setAllYNABDatabases()

2021-10-03T12:34:16.439369+00:00 app[web.1]:   File "/app/helper.py", line 111, in setAllYNABDatabases

2021-10-03T12:34:16.439450+00:00 app[web.1]:     budget = classes.YNABBudget(response.json()["data"]["budget"])

2021-10-03T12:34:16.439451+00:00 app[web.1]:   File "/app/classes.py", line 241, in __init__

2021-10-03T12:34:16.439550+00:00 app[web.1]:     self.setPayeeCategoryDatabase()

2021-10-03T12:34:16.439552+00:00 app[web.1]:   File "/app/classes.py", line 266, in setPayeeCategoryDatabase

2021-10-03T12:34:16.439642+00:00 app[web.1]:     response = requests.get(

2021-10-03T12:34:16.439659+00:00 app[web.1]: KeyError: 'data'

**Comment by CyberiumShadow** [comment]
Having a further look by printing each transaction to console, it looks like it's crashing on a transaction that was split between multiple categories.

I'm assuming this might be why?

**Comment by BrodieSutherland** [comment]
Hey, thanks for reaching out! Yep, you've found the issue correctly, I've just updated the repo with a fix for this here: #3, unfortunately the YNAB full budget retrieval payload doesn't include any info in the transactions to differentiate between the Split transactions and Standard ones, so I've had to include another call to the YNAB API, but it's worked when I've tested locally.

Could you try restarting the dynos and let me know how that goes?

**Comment by CyberiumShadow** [comment]
```
UP TRANSACTION:

Date: 2021-10-09T12:39:25+11:00

Value: 11.42

Payee: Transfer from Spending

Account: 💰 Savings

Message: 

Status: SETTLED

Is Internal: False

11.42 paid to Transfer from Spending at 2021-10-09T12:39:25+11:00
[09/Oct/2021 01:39:27] "[37mPOST /up_webhook HTTP/1.1[0m" 200 -

UP TRANSACTION:

Date: 2021-10-09T12:39:24+11:00
Value: -11.42

Payee: Transfer to Savings

Account: Spending

Message: 

Status: SETTLED

Is Internal: True

[2021-10-09 01:39:27,273] ERROR in app: Exception on /up_webhook [POST]

Traceback (most recent call last):

File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 2447, in wsgi_app

response = self.full_dispatch_request()

File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1952, in full_dispatch_request

rv = self.handle_user_exception(e)

File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1821, in handle_user_exception

reraise(exc_type, exc_value, tb)

File "/app/.heroku/python/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise

raise value

File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1950, in full_dispatch_request

rv = self.dispatch_request()

File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1936, in dispatch_request

return self.view_functions[rule.endpoint](**req.view_args)

File "/app/main.py", line 17, in respond

outcome = helper.handleWebhookEvent(eventPayload)

File "/app/helper.py", line 22, in handleWebhookEvent

event.convertTransaction()

File "/app/classes.py", line 46, in convertTransaction

self.ynabTransaction = YNABTransaction(upTransaction=self.transaction)

File "/app/classes.py", line 166, in __init__

self.payeeId = helper.getVariableFromShelf(

AttributeError: 'NoneType' object has no attribute 'transferId'

[09/Oct/2021 01:39:27] "[35m[1mPOST /up_webhook HTTP/1.1[0m" 500 -

**Comment by BrodieSutherland** [comment]
🤔 this is really strange, maybe I've made an incorrect assumption that all Transactional accounts are called `Spending`?

If possible, could you run the `List Accounts` curl command [here](https://developer.up.com.au/#get_accounts) and confirm that the `TRANSACTIONAL` account's display name is `Spending`?

**Comment by CyberiumShadow** [comment]
Ah my bad, looks like it was user error, I omitted the money bag emoji from YNAB.

Tried again and the transaction gets recorded on the Savings account in YNAB now but as a blank transaction.
![image](https://user-images.githubusercontent.com/4259904/136640525-bfc0eb55-43c6-40ed-a982-f266afe1aec0.png)

I'm presuming it's supposed to automatically fill the payee with a Transfer to: Spending?

**Comment by CyberiumShadow** [comment]
I can also confirm that the Transactional account is called Spending

**Comment by BrodieSutherland** [comment]
Oh excellent! Yeah that category thing is an issue I fixed just before as well, I had some dumbass string handling i fixed up, if you pull down and redeploy you should be gucci 👌

**Comment by CyberiumShadow** [comment]
Alrighty, Internal Transfers between Spending and Savers works now.

What about "Covers"? That seems to be crashing out now as well or are those not supported?

```
--------------------------------------------------------
[2021-10-09 03:54:58,669] ERROR in app: Exception on /up_webhook [POST]

Traceback (most recent call last):

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 2447, in wsgi_app

     response = self.full_dispatch_request()

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1952, in full_dispatch_request

     rv = self.handle_user_exception(e)

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1821, in handle_user_exception

     reraise(exc_type, exc_value, tb)

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise

     raise value

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1950, in full_dispatch_request

     rv = self.dispatch_request()

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1936, in dispatch_request

     return self.view_functions[rule.endpoint](**req.view_args)

   File "/app/main.py", line 17, in respond

     outcome = helper.handleWebhookEvent(eventPayload)

   File "/app/helper.py", line 17, in handleWebhookEvent

     event.getTransaction()

   File "/app/classes.py", line 34, in getTransaction

     self.transaction = UpTransaction(response.json()["data"])

   File "/app/classes.py", line 76, in __init__

     print(self)

   File "/app/classes.py", line 80, in __str__

     "UP TRANSACTION:\n"

 TypeError: can only concatenate str (not "NoneType") to str

--------------------------------------------------------
[2021-10-09 03:54:59,344] ERROR in app: Exception on /up_webhook [POST]

 Traceback (most recent call last):

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 2447, in wsgi_app

     response = self.full_dispatch_request()

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1952, in full_dispatch_request

     rv = self.handle_user_exception(e)

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1821, in handle_user_exception

     reraise(exc_type, exc_value, tb)

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/_compat.py", line 39, in reraise

     raise value

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1950, in full_dispatch_request

     rv = self.dispatch_request()

   File "/app/.heroku/python/lib/python3.9/site-packages/flask/app.py", line 1936, in dispatch_request

     return self.view_functions[rule.endpoint](**req.view_args)

   File "/app/main.py", line 17, in respond

     outcome = helper.handleWebhookEvent(eventPayload)

   File "/app/helper.py", line 17, in handleWebhookEvent

     event.getTransaction()

   File "/app/classes.py", line 34, in getTransaction

     self.transaction = UpTransaction(response.json()["data"])

   File "/app/classes.py", line 76, in __init__

     print(self)

   File "/app/classes.py", line 80, in __str__

     "UP TRANSACTION:\n"

 TypeError: can only concatenate str (not "NoneType") to str

**Comment by BrodieSutherland** [comment]
They should be supported, looks like a bug with the debug stuff i cobbled together, if you turn off debug mode this particular one should be resolved, I'll update this issue once I've fixed this one though

**Comment by CyberiumShadow** [comment]
Alrighty, looks to be all good now. 

I'll probably have to give it a few days to test out other functionality in the Up app such as Split Payments (the % based splitting between Transactional and Savers)

I've yet to see if Round Ups work too

**Comment by BrodieSutherland** [comment]
In my experience roundups and split payments are both totally fine, but keen to hear your experience with them!

I'll be honest, I'm actually not sure. I've got a hobby dyno set up (I think it's $7 USD a month?) for 100% uptime, but if I had to guess I would say it's unlikely to work correctly. Once it comes out of hibernation I think it'll be too busy refreshing the category/account databases to handle the webhook event, but give it a try and see what happens!
